bin/plotting/pycbc_page_coinc_snrchi

```diff
@@ -108,10 +126,24 @@ if 'optimal_snr_{}'.format(ifo) in f['injections']:
     coloring['optimal_snr'] = (opt_snr, 'Optimal SNR', colors.LogNorm())
 
 f = h5py.File(args.single_injection_file, 'r')[ifo]
-tid = inj_tids[ifo][:]
-if len(tid):
-    inj_snr = f['snr'][:][tid]
-    inj_chisq = get_chisq_from_file_choice(f, args.chisq_choice)[tid]
+if len(inj_tid):
+    inj_trigs = SingleDetTriggers(
+        args.single_injection_file,
+        ifo,
+        premask=inj_tid,
+    )
+    inj_snr = inj_trigs.snr
+    inj_chisq = get_chisq_from_file_choice(inj_trigs, args.chisq_choice)
+
+    # Again, the SingleDetTriggers will have discarded order etc. of tids,
+    # so we need to do index juggling
+    _, inj_tid_idx_inverse = numpy.unique(
+        inj_tid,
+        return_inverse=True
+    )
```

Comment:
This is nice. Can you comment on the efficiency of this method compared to:
```
    mask = numpy.zeros(len(f[f'{ifo}/snr']), dtype=bool)
    mask[inj_tid] = True
    tid_locations = numpy.where(mask)[0]
    inj_tid_idx_inverse = numpy.searchsorted(
        tid_locations,
        inj_tid
    )
 ```
 
 which was used before, and is (I think) in use elsewhere in PyCBC for the same thing.

Reply:
(Specifically, what is compute cost in Big-O notation of both methods, and memory usage of both methods).

Reply:
I will calculate this tomorrow (I think), but I just wanted to note that the method you have given will not work for negative numbers. The unique method will retain negative values, which is why I used that instead, given that we often use -1 to signify 'invalid' values 

I thought that the unique version was a bit more readable as well

Reply:
I agree that the new version is more readable. But if this is O(N^2) vs O(N log N) in terms of CPU usage we shouldn't do it! (I think `searchsorted` is O(N log N) but not 100% sure!)

Reply:
Timing both methods shows this:
```
In [8]: def bool_where_searchsorted(tids, snr_size):
   ...:     mask = np.zeros(snr_size, dtype=bool)
   ...:     mask[tids] = True
   ...:     tid_locations = np.where(mask)[0]
   ...:     inj_tid_idx_inverse = np.searchsorted(
   ...:         tid_locations,
   ...:         tids,
   ...:     )
   ...:     return inj_tid_idx_inverse


In [14]: def unique_inverse(tids, snr_size):
    ...:     _, inj_tid_idx_inverse = np.unique(
    ...:         tids,
    ...:         return_inverse=True
    ...:     )
    ...:     return inj_tid_idx_inverse

In [30]: %timeit unique_inverse(np.random.randint(size=100, low=0, high=1000), 1000)
22.9 µs ± 2.55 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [31]: %timeit bool_where_searchsorted(np.random.randint(size=100, low=0, high=1000), 1000)
14 µs ± 557 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [32]: %timeit unique_inverse(np.random.randint(size=100, low=0, high=100000), 100000)
23.5 µs ± 1.83 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [33]: %timeit bool_where_searchsorted(np.random.randint(size=100, low=0, high=100000), 100000)
28.5 µs ± 729 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [34]: %timeit bool_where_searchsorted(np.random.randint(size=1000, low=0, high=1000000), 1000000)
232 µs ± 6.64 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [35]: %timeit unique_inverse(np.random.randint(size=1000, low=0, high=1000000), 1000000)
69.5 µs ± 797 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [36]: %timeit unique_inverse(np.random.randint(size=100, low=0, high=10000), 10000)
23.4 µs ± 1.09 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [37]: %timeit unique_inverse(np.random.randint(size=100, low=0, high=100000), 100000)
22.9 µs ± 816 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Which I think makes sense - the boolean mask/where/searchsorted method is quicker for small reference arrays. But the unique inverse method doesn't depend on the size of the reference array, whereas the searchsorted method does (see In[36], In[37] and In[30] vs In[31] and In[33]).

Looking at this test for a typical O4 chunk, we would have 1e6 background events, and 1.6e9 values in the trigger merge file - in this case the unique inverse function outperforms the boolean mask/where/searchsorted array by a factor of 6
```
In [5]: %timeit bool_where_searchsorted(np.random.randint(size=int(1e6), low=0, high=int(1.6e9)), int
   ...: (1.6e9))
1.54 s ± 50.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit unique_inverse(np.random.randint(size=int(1e6), low=0, high=int(1.6e9)), int(1.6e9))
250 ms ± 111 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Though either is still only called relatively few times within the code, and takes a few seconds so shouldn't matter too much.

For the 1.6e9 / 1e6 example from before, the memory profiles are:
```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    15     67.3 MiB     67.3 MiB           1   @profile
    16                                         def unique_inverse(tids, snr_size):
    17     84.1 MiB     16.8 MiB           2       _, inj_tid_idx_inverse = np.unique(
    18     67.3 MiB      0.0 MiB           1           tids,
    19     67.3 MiB      0.0 MiB           1           return_inverse=True
    20                                             )
    21     84.1 MiB      0.0 MiB           1       return inj_tid_idx_inverse


Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
     4     76.4 MiB     76.4 MiB           1   @profile
     5                                         def bool_where_searchsorted(tids, snr_size):
     6     76.4 MiB      0.0 MiB           1       mask = np.zeros(snr_size, dtype=bool)
     7   1483.3 MiB   1406.9 MiB           1       mask[tids] = True
     8   1483.5 MiB      0.2 MiB           1       tid_locations = np.where(mask)[0]
     9   1490.1 MiB      6.6 MiB           2       inj_tid_idx_inverse = np.searchsorted(
    10   1483.5 MiB      0.0 MiB           1           tid_locations,
    11   1483.5 MiB      0.0 MiB           1           tids,
    12                                             )
    13   1490.1 MiB      0.0 MiB           1       return inj_tid_idx_inverse
```

As expected, setting the boolean mask to True is the real memory hog of the searchsorted method, but the unique method is (to me) surprisingly efficient with memory

Reply:
Okay, it does seem that both have O(N log N) complexity. I have a hunch that this could be done with O(N) complexity (pandas has a `unique` function that claims this, but I don't think it returns the inverse).

... So happy to use numpy.unique over numpy.searchsorted for this kind of operation .... You might want to see how often searchsorted is used in PyCBC at the moment!